CRM/Utils/Hook.php

```diff
@@ -1635,6 +1635,23 @@ public static function disable() {
     );
   }
 
+  /**
+   * Alter redirect.
+   *
+   * This hook is called when the browser is being re-directed and allows the url
+   * to be altered.
+   *
+   * @param string $url
```

Comment:
Not a hard blocker, but just to verbalize -- wouldn't it scale-up better (to support many different listeners) to pass the URL in `array` or `object` format (rather than string format)? i.e. it would avoid redundant parsing/encoding logic?

Reply:
I'm not 100% sure what you mean here - the url is a string in the redirect function - are suggesting unpacking it in that function to pass out as an array & recompiling it afterwards?

Reply:
Maybe replace line 443 in CRM/Utils/System.php with a more sophisticated conversion of the url into an object (e.g. $parsed_url = parse_url($url)), and then pass that into alterRedirect. Then any alter functions don't have to reparse it, since I would assume the first order of any alter would be to do things like see if it's sending you offsite, or going to a particular path. Also, not clear of the value of returning something from the alterRedirect function since the url is going to be modified by reference, and the caller on 445 isn't paying attention to the return value.

Reply:
@adixon Yeah, `parse_url()` should do a pretty good job. Might also add `parse_str()` so that the query-parameters are an array:

```php
$parsedUrl = parse_url($url);
if (isset($parsedUrl['query'])) {
  parse_str($parsedUrl['query'], $parsedUrl['query']);
}
```

(Example inputs/outputs: https://gist.github.com/totten/82918724be1293ffed4e4f7f87d01706 )

It takes a bit more code to put it back into string form. [Here's an example of re-encoding the URL](https://stackoverflow.com/a/35207936/4195300).

Re: `return` -- I think that's just pro-forma. Most hook-stubs have the `return` statement, and the IDE really wants you want to document it, so you see a lot of snippets like this:

```
   * @return null
   *   the return value is ignored
```

Reply:
So to be clear @totten - you agree with @adixon that it would be better to call parse url & then re-encode the string after the hook?

I'm happy to do that - I just want to be sure there is consensus on it

Reply:
@eileenmcnaughton Yup.

Reply:
At first I was skeptical of parsing and re-encoding all redirected URLs all of the time, but realistically *any* consumer of this hook is going to be doing just that in order to decide whether to alter the redirect.